src/main/groovy/com/github/jengelman/gradle/plugins/shadow/ShadowExtension.groovy

```diff
@@ -1,33 +1,38 @@
 package com.github.jengelman.gradle.plugins.shadow
 
+import com.github.jengelman.gradle.plugins.shadow.tasks.ShadowJar
 import org.gradle.api.Project
 import org.gradle.api.artifacts.ProjectDependency
 import org.gradle.api.artifacts.SelfResolvingDependency
+import org.gradle.api.file.RegularFile
+import org.gradle.api.provider.Provider
 import org.gradle.api.publish.maven.MavenPom
 import org.gradle.api.publish.maven.MavenPublication
-import org.gradle.util.GradleVersion
 
 class ShadowExtension {
+    private final Provider<Provider<RegularFile>> archiveFile
+    private final Provider<List<Dep>> allDependencies
 
-    void component(Project project, MavenPublication publication) {
-
-        if (GradleVersion.current() >= GradleVersion.version("6.6")) {
-            publication.artifact(project.tasks.named("shadowJar"))
-        } else {
-            publication.artifact(project.tasks.shadowJar)
-        }
-
-        final def allDependencies = project.provider {
-            project.configurations.shadow.allDependencies.collect {
-                if ((it instanceof ProjectDependency) || ! (it instanceof SelfResolvingDependency)) {
+    ShadowExtension(Project project) {
+        archiveFile = project.provider { project.tasks.withType(ShadowJar).getByName("shadowJar").archiveFile }
```

Comment:
:lipstick: It would be cleaner if the `archiveFile` field was a `Provider<RegularFile>` instead.
Something like the following should do:
```groovy
if (GradleVersion.current() >= GradleVersion.version("6.6")) {
    archiveFile = project.tasks.withType(ShadowJar).named('shadowJar').flatMap { it.archiveFile }
} else {
    archiveFile = project.tasks.withType(ShadowJar).getByName('shadowJar').archiveFile
}
```

Reply:
I tried, it's too eager, see 27eb364211a6d73f87c7c9bfede8fcdcd239bc35

Reply:
I tried `project.tasks.named('shadowJar').map { it.archiveFile }` too in days ago, both failed, so I just wrapped it with `provider` again.